runtime-modules/content-directory/src/lib.rs

```diff
@@ -0,0 +1,1418 @@
+        &self.class_permissions
+    }
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct Entity<T: Trait> {
+    #[cfg_attr(feature = "std", serde(skip))]
+    pub entity_permission: EntityPermissions<T>,
+
+    /// The class id of this entity.
+    pub class_id: T::ClassId,
+
+    /// What schemas under which this entity of a class is available, think
+    /// v.2.0 Person schema for John, v3.0 Person schema for John
+    /// Unlikely to be more than roughly 20ish, assuming schemas for a given class eventually stableize, or that very old schema are eventually removed.
+    pub supported_schemas: BTreeSet<SchemaId>, // indices of schema in corresponding class
+
+    /// Values for properties on class that are used by some schema used by this entity!
+    /// Length is no more than Class.properties.
+    pub values: BTreeMap<PropertyId, PropertyValue<T>>,
+    // pub deleted: bool
+    pub reference_count: u32,
+}
+
+impl<T: Trait> Default for Entity<T> {
+    fn default() -> Self {
+        Self {
+            entity_permission: EntityPermissions::<T>::default(),
+            class_id: T::ClassId::default(),
+            supported_schemas: BTreeSet::new(),
+            values: BTreeMap::new(),
+            reference_count: 0,
+        }
+    }
+}
+
+impl<T: Trait> Entity<T> {
+    fn new(
+        controller: Option<EntityController<T>>,
+        class_id: T::ClassId,
+        supported_schemas: BTreeSet<SchemaId>,
+        values: BTreeMap<PropertyId, PropertyValue<T>>,
+    ) -> Self {
+        Self {
+            entity_permission: EntityPermissions::<T>::default_with_controller(controller),
+            class_id,
+            supported_schemas,
+            values,
+            reference_count: 0,
+        }
+    }
+
+    fn get_permissions_mut(&mut self) -> &mut EntityPermissions<T> {
+        &mut self.entity_permission
+    }
+
+    fn get_permissions(&self) -> &EntityPermissions<T> {
+        &self.entity_permission
+    }
+}
+
+// Shortcuts for faster readability of match expression:
+use PropertyType as PT;
+use PropertyValue as PV;
+
+decl_storage! {
+    trait Store for Module<T: Trait> as ContentDirectory {
+        pub ClassById get(class_by_id) config(): linked_map T::ClassId => Class<T>;
+
+        pub EntityById get(entity_by_id) config(): map T::EntityId => Entity<T>;
+
+        pub NextClassId get(next_class_id) config(): T::ClassId;
+
+        pub NextEntityId get(next_entity_id) config(): T::EntityId;
+
+        /// Groups who's actors can create entities of class.
+        pub CanCreateEntitiesOfClass get(can_create_entities_of_class): double_map hasher(blake2_128) T::ClassId, blake2_128(T::GroupId) => ();
+
+        /// Groups who's actors can act as entity maintainers.
+        pub EntityMaintainers get(entity_maintainers): double_map hasher(blake2_128) T::EntityId, blake2_128(T::GroupId) => ();
```

Comment:
Just wanted to note that maintainers are associated with classes, they are the maintainers of _all_ entities of a given class. This associates them with individual entities, that is not correct. Perhaps another name would have avoided the misunderstanding, like `EntityMaintainersForClass`. Just making a note of it, but this field goes away with the new actor/authenticaiton model anyway.

Reply:
Done https://github.com/Joystream/joystream/pull/396/commits/07cbb66182a5ce03280a91dc91679f5418c4bc5c